File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
var validPath = function(n, edges, source, destination) {
 let visited = new Set();
  const graph = {}
  let answer = false
  
  for(const [from, to] of edges){
    if(graph[from]){
      graph[from].push(to)
    } else {
      graph[from] = [to]
    }
    if(graph[to]){
      graph[to].push(from)
    } else {
      graph[to] = [from]
    }
  }
  const bfs = () => {
    const queue = [source]
    
    while(queue.length){
      const node = queue.shift()
      if(node === destination) {
        answer = true
        return 
      }
      if(graph[node]){
        for(const nextNode of graph[node]){
           if(!visited.has(nextNode)) {
              visited.add(nextNode)  
               queue.push(nextNode) 
           }
        }
      }
    }
  }
  bfs()
  return answer
};

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        Map<Integer, List<Integer>> graph = new HashMap();
        
        for(int[] e : edges) {
            int from = e[0];
            int to = e[1];
            
            graph.computeIfAbsent(from, k -> new ArrayList<>()).add(to);
            graph.computeIfAbsent(to, k -> new ArrayList<>()).add(from);
        }
        
        boolean[] visited = new boolean[n + 1];
        
        Queue<Integer> queue = new LinkedList<>();
        queue.add(source);
        
        visited[source] = true;
        
        while(!queue.isEmpty()) {
            int node = queue.poll();
            
            if(node == destination) return true;
            
            if(graph.containsKey(node)){
                for(int neighbor : graph.get(node)){
                    if(!visited[neighbor]){
                        visited[neighbor] = true;
                        queue.add(neighbor);               
                    }
                }
            }
        }
        
        return false;
    }
}